test/NuGet.Core.Tests/NuGet.Build.Tasks.Console.Test/ProgramTests.cs

```diff
@@ -0,0 +1,204 @@
+
+            stream.Position = 0;
+
+            using var reader = new BinaryReader(stream, Encoding.UTF8, leaveOpen: true);
+
+            bool result = Program.TryDeserializeGlobalProperties(errorWriter, reader, out Dictionary<string, string> actualGlobalProperties);
+
+            actualGlobalProperties.Should().BeEquivalentTo(expectedGlobalProperties);
+        }
+
+        /// <summary>
+        /// Verifies that <see cref="Program.TryDeserializeGlobalProperties(TextWriter, BinaryReader, out Dictionary{string, string})" /> returns <see langword="false" /> and logs an error if the stream contains an unsupported integer.
+        /// </summary>
+        /// <param name="bytes">An array of bytes to use as the stream.</param>
+        [Theory]
+        [InlineData(new byte[] { 0x00, 0x00, 0x00, 0xFF })] // 4 byte integer that is negative
+        [InlineData(new byte[] { 0xFF, 0xFF, 0xFF, 0x7F })] // 4 byte integer that is too big
+        public void TryDeserializeGlobalProperties_WhenInvalidDictionaryLength_ReturnsFalse(byte[] bytes)
+        {
+            using var stream = GetStreamWithBytes(bytes);
+
+            int expectedLength = BitConverter.ToInt32(bytes, bytes.Length - 4);
+
+            VerifyTryDeserializeGlobalPropertiesError(stream, Strings.Error_StaticGraphRestoreArgumentsParsingFailedUnexpectedIntegerValue, expectedLength);
+        }
+
+        /// <summary>
+        /// Verifies that <see cref="Program.TryDeserializeGlobalProperties(TextWriter, BinaryReader, out Dictionary{string, string})" /> returns <see langword="false" /> and logs an error if reading the stream causes an exception to be thrown.
+        /// </summary>
+        /// <param name="throwOnReadCount">Indicates for which call to Stream.Read() should throw, the first, second, or third.</param>
+        /// <param name="bytes">An array of bytes to use as the stream.</param>
+        [Theory]
+        [InlineData(1)] // Throw on first Read()
+        [InlineData(2)] // Throw on second Read()
+        public void TryDeserializeGlobalProperties_WhenStreamReadThrows_ReturnsFalse(int throwOnReadCount)
+        {
+            var expectedGlobalProperties = new Dictionary<string, string>
+            {
+                ["Property1"] = "Value1",
+                ["Property2"] = "  Value2  "
+            };
+
+            Exception exception = new InvalidOperationException();
+
+            using var stream = new StreamThatThrowsAnException(BitConverter.GetBytes(1), throwOnReadCount, exception);
+
+            VerifyTryDeserializeGlobalPropertiesErrorStartsWith(stream, Strings.Error_StaticGraphRestoreArgumentsParsingFailedExceptionReadingStream, exception.Message, string.Empty);
+        }
+
+        private static Stream GetStreamWithBytes(params byte[] bytes)
+        {
+            MemoryStream stream = new MemoryStream();
+
+            using BinaryWriter writer = new BinaryWriter(stream, Encoding.UTF8, leaveOpen: true);
```

Comment:
Search for 3 instances of `new BinaryWriter(` in tests.  Should they use a no-BOM-emitting UTF8 encoder?  Why or why not?

Reply:
BinaryWriter never emits a BOM and only uses the encoding to determine the character width.